Feat/add browser based tests

chelseapinka opened the pull request:
# Adding Browser based E2E tests

Added two test cases
 - Ensure that VCs can be issued and then revoked
 - Ensure that VCs will not be issued without the required minimum credentials. 
 - Added named exports for modules that were missing them

# Checklist

- [x] All acceptance criteria are met.
- [x] Relevant documentation, if any, has been written/updated.
- [x] The changelog has been updated, if applicable.
- [x] New functions/types have been exported in `index.ts`, if applicable.
- [x] New modules (i.e. new `.ts` files) are listed in the `exports` field in `package.json`, if applicable.
- [x] New modules (i.e. new `.ts` files) are listed in the `typedocOptions.entryPoints` field in `tsconfig.json`, if applicable.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

chelseapinka committed:
feat(browser-e2e-tests): Adding base configuration of playwright to run browser scenarios.

chelseapinka committed:
feat(browser-e2e-tests): Creating and cleaning up resource button components for testing runs.

chelseapinka committed:
feat(browser-e2e-tests): Adding invalid credentials test case

chelseapinka committed:
feat(browser-e2e-tests): Fixing vcProvider field to be updated from Envrionement variables on component load

chelseapinka committed:
feat(browser-e2e-tests): Adding internal-test-env package version bump

chelseapinka committed:
feat(browser-e2e-tests): Fix missing variable for handleIssue call on invalid vc

chelseapinka committed:
feat(browser-e2e-tests): Click close for unhandled error popup

chelseapinka committed:
feat(browser-e2e-tests): Fix linting issues

chelseapinka committed:
feat(browser-e2e-tests): Updated lintin extension rules

chelseapinka committed:
feat(browser-e2e-tests): Insert fill of vc issuser endpoint

ThisIsMissEm commented:
Try moving the waitFor's before the interactions (timing here is tricky, but the click might be happening before the waitFor's — this is why in other tests we assert on DOM elements being added or removed).

also the fill on line 41 should probably be above / outside of that promise.all block

chelseapinka committed:
feat(browser-e2e-tests): Switch waitFor and click interaction

chelseapinka committed:
feat(browser-e2e-tests): Switch waitFor and click interaction on close of error modal

chelseapinka committed:
feat(browser-e2e-tests): Change test app deps to devDeps

chelseapinka committed:
feat(browser-e2e-tests): Manually construct the URL

solid-akb committed:
iri test

solid-akb committed:
remove console.logs

solid-akb committed:
try pulling the provider var directly from state

chelseapinka committed:
feat(browser-e2e-tests): Add env var to github workflow

chelseapinka committed:
feat(browser-e2e-tests): build URL string

chelseapinka committed:
feat(browser-e2e-tests): update to getBrowserTestingEnv

chelseapinka committed:
feat(browser-e2e-tests): lint for test-app config updated

chelseapinka committed:
feat(browser-e2e-tests): update test-apps linting packages

chelseapinka committed:
feat(browser-e2e-tests): remove next eslint configs

chelseapinka committed:
feat(browser-e2e-tests): don't worry about linting the test app

chelseapinka committed:
feat(browser-e2e-tests): add optional depenedency for dependency warning

chelseapinka committed:
feat(browser-e2e-tests): add ESS PodSpaces for browser testing

chelseapinka committed:
docs: Improve generation of documentation
This is a fairly significant change, but fixes a bunch of things with our vercel-built documentation pages as well as improving the developer experience when working on documentation.

Scripts changes:
- npm run build-api-docs => npm docs:build
- npm run build-api-docs-preview-site => npm docs:build
- Added docs:clean (removes all generated files, typically you won't need this)
- Added docs:install (simplifies installation)

Improvements:
- Updated docs README with new instructions
- Moved docs-assets out of the build directory so it can be cached between builds, allowing faster docs generation (run docs:clean to update it)
- Added makefile command for generating the dist/ output
- Added warning if you try to run the makefile without sphinx-build being available, instead of crashing
- Pinned myst-parser and Sphinx dependency versions, as these broke the docs build due to myst-parser not yet having a major version.

chelseapinka committed:
docs:add missing types from solid-client for building docs

chelseapinka committed:
chore: ensure default exports are also exported as named exports

chelseapinka committed:
docs: improve alpha banner text

chelseapinka committed:
chore: remove outdated jest config comment

chelseapinka committed:
chore: added changelog updates

NSeydoux commented on .eslintignore:
```diff
@@ -2,3 +2,4 @@ jest.config.ts
 license-header.js
 rollup.config.js
 tsconfig.json
+e2e/browser/test-app
```
Why add the test app to the unlinted files? Does it prevent conflicts with its own linter config?

NSeydoux commented on e2e/browser/test-app/components/appContent/index.tsx:
```diff
@@ -0,0 +1,136 @@
+// This is the content of the file uploaded manually at SHARED_FILE_IRI.
+const DEFAULT_ISSUER = "https://login.inrupt.com/";
+const REDIRECT_URL = window.location.href;
+const APP_NAME = "Access Grants browser-based tests app";
```
```suggestion
const APP_NAME = "Verifiable Credentials browser-based tests app";
```

NSeydoux commented on e2e/browser/test-app/components/appContent/index.tsx:
```diff
@@ -0,0 +1,136 @@
+const DEFAULT_ISSUER = "https://login.inrupt.com/";
+const REDIRECT_URL = window.location.href;
+const APP_NAME = "Access Grants browser-based tests app";
+const AccessGrantContainer = ({
```
Even if the VCs here happen to be access grants, I would be in favour of gearing the naming towards VC not to emphasis on the data model, but on its spec-compliant container.

NSeydoux commented on e2e/browser/test-app/styles/Home.module.css:
```diff
@@ -0,0 +1,129 @@
+.container {
```
I think it's fine to strip the test app of all styling

NSeydoux commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -0,0 +1,117 @@
+  // eslint-disable-next-line playwright/no-conditional-in-test
+  await page.fill("input[data-testid=vcProvider]", vcProvider || "");
+  // Create the resource
+  await Promise.all([
+    page.click("button[data-testid=create-resource]"),
+    page.waitForRequest((request) => request.method() === "POST"),
+    page.waitForResponse((response) => response.status() === 201),
+  ]);
+  await expect(
+    page.innerText("span[data-testid=resource-iri]")
+  ).resolves.toMatch(/https:\/\/.*\.txt/);
```
I wouldn't bother creating a dedicated resource. It makes sense if testing Access Grant capabilities, i.e. "can I use the issued Access Grant to get access to a resource", but all we want to test here is "can I issue a VC passing some data in and checking that said data made it through". The fact that the VC happens to be an Access Grant is not that important, as the Access Grant features are tested in the dedicated library.

That means this step can be removed, as well as the related components in the test app.

NSeydoux replied:
```suggestion
```

NSeydoux commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -0,0 +1,117 @@
+    page.innerText("pre[data-testid=verifiable-credential]")
+  ).resolves.not.toBe("");
+
+  // TODO add extra validation of claims here?
```
That sounds good, maybe parsing the VC as JSON and checking for some properties?

NSeydoux commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -0,0 +1,117 @@
+    page.innerText("pre[data-testid=verifiable-credential]")
+  ).resolves.not.toBe("");
+
+  // Cleanup the resource
+  await Promise.all([
+    page.click("button[data-testid=delete-resource]"),
+    page.waitForRequest((request) => request.method() === "DELETE"),
+    page.waitForResponse((response) => response.status() === 204),
+  ]);
+  await expect(
+    page.innerText("span[data-testid=resource-iri]")
+  ).resolves.toMatch("");
```
```suggestion
```
See above comment about VC vs Access Grants

NSeydoux commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -0,0 +1,117 @@
+  auth,
+}) => {
+  await auth.login({ allow: true });
+  // Create the resource
+  await Promise.all([
+    page.click("button[data-testid=create-resource]"),
+    page.waitForRequest((request) => request.method() === "POST"),
+    page.waitForResponse((response) => response.status() === 201),
+    // eslint-disable-next-line playwright/no-conditional-in-test
+    page.fill("input[data-testid=vcProvider]", vcProvider || ""),
+  ]);
+  await expect(
+    page.innerText("span[data-testid=resource-iri]")
+  ).resolves.toMatch(/https:\/\/.*\.txt/);
```
```suggestion
```
Ditto

NSeydoux commented on e2e/browser/test/e2e.playwright.ts:
```diff
@@ -0,0 +1,117 @@
+    page.innerText("pre[data-testid=verifiable-credential]")
+  ).resolves.toBe("");
+
+  // Cleanup the resource
+  await Promise.all([
+    page.click("button[data-testid=delete-resource]"),
+    page.waitForRequest((request) => request.method() === "DELETE"),
+    page.waitForResponse((response) => response.status() === 204),
+  ]);
+  await expect(
+    page.innerText("span[data-testid=resource-iri]")
+  ).resolves.toMatch("");
```
```suggestion
```
Ditto

NSeydoux reviewed (commented):
Just a couple of superficial comments for the test app, but that looks good!

ThisIsMissEm reviewed (commented):
I read @NSeydoux's comments and agree, I'll review prop on Monday or Tuesday

chelseapinka committed:
Update e2e/browser/test-app/components/appContent/index.tsx

Co-authored-by: Zwifi <nseydoux@inrupt.com>

chelseapinka committed:
chore: removed unneeded test-app styles

chelseapinka committed:
chore: formatting typo fix

chelseapinka committed:
fix: add validation to VC strucutre

chelseapinka committed:
fix: remove resource creation at beginning of tests

NSeydoux reviewed (approved):
All good!
